File: mobile/src/main/java/treehou/se/habit/module/RxLifecyclePresenter.java
package treehou.se.habit.module;


import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;

import com.trello.rxlifecycle2.LifecycleTransformer;
import com.trello.rxlifecycle2.OutsideLifecycleException;
import com.trello.rxlifecycle2.RxLifecycle;

import io.reactivex.Observable;
import io.reactivex.functions.Function;
import treehou.se.habit.module.RxPresenterEvent.PresenterEvent;

public class RxLifecyclePresenter {
    private RxLifecyclePresenter() {
        throw new AssertionError("No instances");
    }

    @NonNull
    @CheckResult
    public static <T> LifecycleTransformer<T> bindPresenter(@NonNull final Observable<PresenterEvent> lifecycle) {
        return RxLifecycle.bind(lifecycle, PRESENTER_LIFECYCLE);
    }

    private static final Function<PresenterEvent, PresenterEvent> PRESENTER_LIFECYCLE = lastEvent -> {
        switch (lastEvent) {
            case LOAD:
                return PresenterEvent.UNLOAD;
            case SUBSCRIBE:
                return PresenterEvent.UNSUBSCRIBE;
            case UNSUBSCRIBE:
            case UNLOAD:
                throw new OutsideLifecycleException("Cannot bind to Presenter lifecycle when outside of it.");
            default:
                throw new UnsupportedOperationException("Binding to " + lastEvent + " not yet implemented");
        }
    };

}

File: mobile/src/main/java/treehou/se/habit/dagger/RxLifecyclePresenter.kt
package treehou.se.habit.dagger


import android.support.annotation.CheckResult

import com.trello.rxlifecycle2.LifecycleTransformer
import com.trello.rxlifecycle2.OutsideLifecycleException
import com.trello.rxlifecycle2.RxLifecycle

import io.reactivex.Observable
import io.reactivex.functions.Function
import treehou.se.habit.dagger.RxPresenterEvent.PresenterEvent

class RxLifecyclePresenter private constructor() {
    init {
        throw AssertionError("No instances")
    }

    companion object {

        @CheckResult
        fun <T> bindPresenter(lifecycle: Observable<PresenterEvent>): LifecycleTransformer<T> {
            return RxLifecycle.bind(lifecycle, PRESENTER_LIFECYCLE)
        }

        private val PRESENTER_LIFECYCLE = Function<PresenterEvent, PresenterEvent>{ lastEvent ->
            when (lastEvent) {
                RxPresenterEvent.PresenterEvent.LOAD -> PresenterEvent.UNLOAD
                RxPresenterEvent.PresenterEvent.SUBSCRIBE -> PresenterEvent.UNSUBSCRIBE
                RxPresenterEvent.PresenterEvent.UNSUBSCRIBE, RxPresenterEvent.PresenterEvent.UNLOAD -> throw OutsideLifecycleException("Cannot bind to Presenter lifecycle when outside of it.")
                else -> throw UnsupportedOperationException("Binding to $lastEvent not yet implemented")
            }
        }
    }

}
